File: backend/src/modules/steps/step.repository.js
import { pool } from '../../config/db.js';

// ── Ownership helper ──────────────────────────────────────────────────────

export const findTestCaseByIdAndProject = async (id_test_case, id_project) => {
    const result = await pool.query(
        `SELECT id_test_case
         FROM test_cases
         WHERE id_test_case = $1
           AND id_project   = $2
           AND deleted_at   IS NULL`,
        [id_test_case, id_project]
    );
    return result.rows[0];
};

export const findProjectByIdAndUser = async (id_project, id_user) => {
    const result = await pool.query(
        `SELECT id_project
         FROM projects
         WHERE id_project = $1
           AND id_user    = $2
           AND deleted_at IS NULL`,
        [id_project, id_user]
    );
    return result.rows[0];
};

// ── Steps ─────────────────────────────────────────────────────────────────

export const findStepByNumberAndCase = async (id_test_case, step_number) => {
    const result = await pool.query(
        `SELECT id_step
         FROM test_steps
         WHERE id_test_case = $1 AND step_number = $2`,
        [id_test_case, step_number]
    );
    return result.rows[0];
};

export const createStep = async (id_test_case, stepData) => {
    const result = await pool.query(
        `INSERT INTO test_steps (id_test_case, step_number, action, expected_result)
         VALUES ($1, $2, $3, $4)
         RETURNING id_step, step_number, action, expected_result`,
        [id_test_case, stepData.step_number, stepData.action, stepData.expected_result]
    );
    return result.rows[0];
};

export const getStepsByTestCase = async (id_test_case, limit = 100, offset = 0) => {
    const result = await pool.query(
        `SELECT id_step, step_number, action, expected_result
         FROM test_steps
         WHERE id_test_case = $1
         ORDER BY step_number ASC
         LIMIT $2 OFFSET $3`,
        [id_test_case, limit, offset]
    );
    return result.rows;
};

export const getStepById = async (id_step) => {
    const result = await pool.query(
        `SELECT id_step, id_test_case, step_number, action, expected_result
         FROM test_steps
         WHERE id_step = $1`,
        [id_step]
    );
    return result.rows[0];
};

export const updateStep = async (id_step, stepData) => {
    const result = await pool.query(
        `UPDATE test_steps
         SET step_number = $1, action = $2, expected_result = $3
         WHERE id_step = $4
         RETURNING id_step, step_number, action, expected_result`,
        [stepData.step_number, stepData.action, stepData.expected_result, id_step]
    );
    return result.rows[0];
};

export const patchStep = async (id_step, stepData) => {
    const updates = [];
    const values  = [];
    let   idx     = 1;

    if (stepData.step_number    !== undefined) { updates.push(`step_number = $${idx++}`);    values.push(stepData.step_number);    }
    if (stepData.action         !== undefined) { updates.push(`action = $${idx++}`);         values.push(stepData.action);         }
    if (stepData.expected_result !== undefined){ updates.push(`expected_result = $${idx++}`); values.push(stepData.expected_result); }

    if (updates.length === 0) {
        const error = new Error('No fields provided to update.');
        error.statusCode = 400;
        throw error;
    }

    values.push(id_step);
    const result = await pool.query(
        `UPDATE test_steps SET ${updates.join(', ')} WHERE id_step = $${idx}
         RETURNING id_step, step_number, action, expected_result`,
        values
    );
    return result.rows[0];
};

export const deleteStep = async (id_step) => {
    const result = await pool.query(
        `DELETE FROM test_steps WHERE id_step = $1 RETURNING id_step`,
        [id_step]
    );
    return result.rows[0];
};

File: QALearnBackend/Repositories/StepRepository.cs
using Dapper;
using QALearnAPI.Models.Entities;

namespace QALearnAPI.Repositories;

public class StepRepository(IConfiguration cfg) : DbRepository(cfg)
{
    public async Task<bool> ProjectBelongsToUserAsync(Guid projectId, Guid userId)
    {
        using var c = Conn();
        return await c.ExecuteScalarAsync<int>(
            "SELECT 1 FROM projects WHERE id_project = @P AND id_user = @U AND deleted_at IS NULL",
            new { P = projectId, U = userId }) == 1;
    }

    public async Task<bool> TestCaseBelongsToProjectAsync(Guid tcId, Guid projectId)
    {
        using var c = Conn();
        return await c.ExecuteScalarAsync<int>(
            "SELECT 1 FROM test_cases WHERE id_test_case = @TcId AND id_project = @P AND deleted_at IS NULL",
            new { TcId = tcId, P = projectId }) == 1;
    }

    public async Task<bool> StepNumberExistsAsync(Guid tcId, int stepNumber)
    {
        using var c = Conn();
        return await c.ExecuteScalarAsync<int>(
            "SELECT 1 FROM test_steps WHERE id_test_case = @TcId AND step_number = @Num",
            new { TcId = tcId, Num = stepNumber }) == 1;
    }

    public async Task<StepEntity> CreateAsync(Guid tcId, int stepNumber, string action, string expectedResult)
    {
        using var c = Conn();
        return await c.QueryFirstAsync<StepEntity>(
            @"INSERT INTO test_steps (id_test_case, step_number, action, expected_result)
              VALUES (@TcId, @Num, @Action, @Expected)
              RETURNING id_step, step_number, action, expected_result",
            new { TcId = tcId, Num = stepNumber, Action = action, Expected = expectedResult });
    }

    public async Task<IEnumerable<StepEntity>> GetByTestCaseAsync(Guid tcId, int limit = 100, int offset = 0)
    {
        using var c = Conn();
        return await c.QueryAsync<StepEntity>(
            "SELECT id_step, step_number, action, expected_result FROM test_steps WHERE id_test_case = @TcId ORDER BY step_number LIMIT @L OFFSET @O",
            new { TcId = tcId, L = limit, O = offset });
    }

    public async Task<StepEntity?> GetByIdAsync(Guid stepId)
    {
        using var c = Conn();
        return await c.QueryFirstOrDefaultAsync<StepEntity>(
            "SELECT id_step, step_number, action, expected_result FROM test_steps WHERE id_step = @StepId",
            new { StepId = stepId });
    }

    public async Task<StepEntity?> UpdateAsync(Guid stepId, int stepNumber, string action, string expectedResult)
    {
        using var c = Conn();
        return await c.QueryFirstOrDefaultAsync<StepEntity>(
            @"UPDATE test_steps SET step_number = @Num, action = @Action, expected_result = @Expected
              WHERE id_step = @StepId RETURNING id_step, step_number, action, expected_result",
            new { Num = stepNumber, Action = action, Expected = expectedResult, StepId = stepId });
    }

    public async Task<StepEntity?> PatchAsync(Guid stepId, int? stepNumber, string? action, string? expectedResult)
    {
        var parts  = new List<string>();
        var p      = new DynamicParameters();
        p.Add("StepId", stepId);
        if (stepNumber    != null) { parts.Add("step_number = @Num");      p.Add("Num",      stepNumber); }
        if (action        != null) { parts.Add("action = @Action");         p.Add("Action",   action); }
        if (expectedResult != null) { parts.Add("expected_result = @Exp"); p.Add("Exp",      expectedResult); }
        if (parts.Count == 0) throw new ArgumentException("No fields to update");
        using var c = Conn();
        return await c.QueryFirstOrDefaultAsync<StepEntity>(
            $"UPDATE test_steps SET {string.Join(", ", parts)} WHERE id_step = @StepId RETURNING id_step, step_number, action, expected_result", p);
    }

    public async Task DeleteAsync(Guid stepId)
    {
        using var c = Conn();
        await c.ExecuteAsync("DELETE FROM test_steps WHERE id_step = @StepId", new { StepId = stepId });
    }
}
